src/jit/MemoryInl.h

```diff
@@ -1011,3 +1011,349 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
+    case ByteCode::I64AtomicRmw8XorUOpcode:
+    case ByteCode::I64AtomicRmw16XorUOpcode:
+    case ByteCode::I64AtomicRmw32XorUOpcode: {
+        operation = SLJIT_XOR;
+        break;
+    }
+    case ByteCode::I32AtomicRmwXchgOpcode:
+    case ByteCode::I32AtomicRmw8XchgUOpcode:
+    case ByteCode::I32AtomicRmw16XchgUOpcode:
+    case ByteCode::I64AtomicRmwXchgOpcode:
+    case ByteCode::I64AtomicRmw8XchgUOpcode:
+    case ByteCode::I64AtomicRmw16XchgUOpcode:
+    case ByteCode::I64AtomicRmw32XchgUOpcode: {
+        operation = OP_XCHG;
+        break;
+    }
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+        break;
+    }
+
+    AtomicRmw* rmwOperation = reinterpret_cast<AtomicRmw*>(instr->byteCode());
+    offset = rmwOperation->offset();
+
+    Operand* operands = instr->operands();
+    MemAddress addr(options, instr->requiredReg(0), instr->requiredReg(1), 0);
+
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+    JITArgPair valueArgPair;
+    if ((operationSize & SLJIT_32) || operationSize == SLJIT_MOV32) {
+        addr.loadArg.set(operands + 1);
+
+        if (SLJIT_IS_MEM(addr.loadArg.arg)) {
+            addr.options = MemAddress::LoadInteger;
+        }
+    } else {
+        valueArgPair.set(operands + 1);
+
+        addr.loadArg.arg = valueArgPair.arg1;
+        addr.loadArg.argw = valueArgPair.arg1w;
+
+        if (SLJIT_IS_MEM(valueArgPair.arg1)) {
+            addr.options = MemAddress::LoadInteger;
+        }
+    }
+#else /* !SLJIT_32BIT_ARCHITECTURE */
+    addr.loadArg.set(operands + 1);
+
+    if (SLJIT_IS_MEM(addr.loadArg.arg)) {
+        addr.options = MemAddress::LoadInteger;
+
+        if (operationSize == SLJIT_MOV32 || (operationSize & SLJIT_32)) {
+            addr.options |= MemAddress::Load32;
+        }
+    }
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+
+    addr.check(compiler, operands, offset, size);
+
+    JITArg dst;
+    sljit_s32 srcReg;
+
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+    if ((operationSize & SLJIT_32) || operationSize == SLJIT_MOV32) {
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+        JITArg src(operands + 1);
+        dst = JITArg(operands + 2);
+        srcReg = GET_SOURCE_REG(src.arg, instr->requiredReg(1));
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+    } else {
+        JITArgPair srcPair(operands + 1);
+        JITArgPair dstPair(operands + 2);
+        srcReg = GET_TARGET_REG(srcPair.arg1, instr->requiredReg(1));
+
+        dst.arg = dstPair.arg1;
+        dst.argw = dstPair.arg1w;
+        sljit_emit_op1(compiler, SLJIT_MOV, dstPair.arg2, dstPair.arg2w, SLJIT_IMM, 0);
+    }
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+
+    struct sljit_label* restartOnFailure = sljit_emit_label(compiler);
+    sljit_emit_atomic_load(compiler, operationSize, SLJIT_TMP_DEST_REG, SLJIT_EXTRACT_REG(addr.memArg.arg));
+    sljit_emit_op1(compiler, SLJIT_MOV, dst.arg, dst.argw, SLJIT_TMP_DEST_REG, 0);
```

Comment:
Please move this code after `sljit_set_label`. I mentioned that memory access between atomic load/store is not necessary supported. Now `sljit_emit_atomic_store` keeps the tmp reg value when the operation is successful, so you can safely store it.